src/V8JsiRuntime.cpp

```diff
@@ -28,7 +26,9 @@ using namespace facebook;
 
 namespace v8runtime {
 
-thread_local uint16_t V8Runtime::tls_isolate_usage_counter_ = 0;
+thread_local uint16_t V8Runtime::tls_isolate_usage_counter_{0};
+thread_local std::weak_ptr<V8Runtime::CounterStores>

```

Comment:
>weak_ptr [](start = 18, length = 8)

Ideally we should stop using TLS. In future we want to be able to use The V8 engine in scenarios without a dedicated JS thread. A better approach is to assciate the data with the Runtime, V8 context, or V8 isolate. 
Anyway, there is no mechanism to call destructor on the TLS stored objects. Please reconsider having std::weak_ptr as TLS value. TLS is usually good only for simple values such as numbers or raw pointers.

Reply:
I looked though the code and still do not understand why we cannot associate counters with the Runtime or Isolate instead of TLS.

---
In reply to: [606609468](https://github.com/microsoft/v8-jsi/pull/50#discussion_r606609468) [](ancestors = 606609468)

Reply:
We need to access the counters from the static callback functions that we register to the isolate .. Essentially, what we want is access to "this" runtime pointer in the static callbacks .. The access to counters starts even before the isolate is properly constructed, hence we can't use the isolate data slots .. 

I was also hesistant to add tls data .. I initially had a unique instance id counter to uniquely identify an instance, instead of a thread id .. But it turned out to be a lot more code and uglier. .. As V8 isolates and a thread is 1:1 mapped, I think the tls is fine for our scenario.. Essentialy, we can treat the counters as per isolate counters, not really per-runtime counters ..

We are not really gaining anything by having the weak_ptr .. even a raw pointer in tls with the shared_ptr member for lifetime manager suffice for us .. But, i made this change to weak_ptr just before the CR to make the code look nicer :) the naked pointer looks bad .. But as you suggested this, i'll revert the tls entry to a nake pointer.